hrjobcontract/CRM/Hrjobcontract/Upgrader.php

```diff
@@ -1065,41 +1065,91 @@ public function createInsurancePlanTypes() {
+   * 
+   * @throws Exception
    */
   function upgrade_1026() {
     $messages = [];
     $pensionsTableName = CRM_Hrjobcontract_BAO_HRJobPension::getTableName();
 
-    $this->createPensionProviderType($messages);
-    $this->replacePensionProviders($pensionsTableName, $messages);
-    $this->removePensionTypeOptionGroup($messages);
-
-    CRM_Core_DAO::executeQuery("
-      ALTER TABLE {$pensionsTableName} 
-      MODIFY pension_type INT(10) unsigned,
-      ADD CONSTRAINT pension_type_contact_id_fk
-      FOREIGN KEY(pension_type)
-      REFERENCES civicrm_contact(id)
-      ON DELETE SET NULL ON UPDATE CASCADE;
-    ");
+    $tx = new CRM_Core_Transaction();
+    try {
+      $this->createPensionProviderType($messages);
+      $this->replacePensionProviders($pensionsTableName, $messages);
+      $this->removePensionTypeOptionGroup($messages);
+      $this->alterPensionsTable($pensionsTableName, $messages);
+    } catch (Exception $ex) {
+      $tx->rollback();
+      throw new Exception('Cannot do upgrade - ' . $ex->getMessage());
+    }
 
     if (!empty($messages) && function_exists('drupal_set_message')) {
       foreach ($messages as $message) {
         drupal_set_message($message);
       }
     }
 
-    return true;
+    return false;
+  }
+
+  /**
+   * Alters pension_type column to be an unsigned integer and adds a foreign
+   * key referencing civicrm_contact.
+   * 
+   * @param type $pensionsTableName
+   *   Name of pension data table
+   * @param type $messages
+   *   Array of messages for salvageable errors found
+   * 
+   * @throws Exception
+   */
+  function alterPensionsTable($pensionsTableName, &$messages) {
+    try {
+      CRM_Core_DAO::executeQuery("
+        ALTER TABLE $pensionsTableName
+        MODIFY pension_type INT(10) unsigned,
+        ADD CONSTRAINT pension_type_contact_id_fk
+        FOREIGN KEY(pension_type)
+        REFERENCES civicrm_contact(id)
+        ON DELETE SET NULL ON UPDATE CASCADE;
+      ");
+    } catch (Exception $e) {
+      $civi_db_settings = parse_url(CIVICRM_DSN);
+      $civi_db_name = trim($civi_db_settings['path'], '/');
+
+      $fkCheck = CRM_Core_DAO::executeQuery($q = "
+        SELECT *
+        FROM information_schema.REFERENTIAL_CONSTRAINTS
+        WHERE CONSTRAINT_SCHEMA = '$civi_db_name'
+        AND TABLE_NAME = '$pensionsTableName'
+        AND REFERENCED_TABLE_NAME = 'civicrm_contact'
+        AND CONSTRAINT_NAME = 'pension_type_contact_id_fk'
+      ");
+      
+      if ($fkCheck->N == 0) {
+        throw new Exception("Error updating $pensionsTableName table: " . $e->getMessage());
+      }
+
+      $pensionTypeColumnCheck = CRM_Core_DAO::executeQuery("DESC $pensionsTableName");
+      while ($pensionTypeColumnCheck->fetch()) {
+        if ($pensionTypeColumnCheck->Field == 'pension_type' && $pensionTypeColumnCheck->Type != 'int(10) unsigned') {
+          throw new Exception("Error updating $pensionsTableName table: " . $e->getMessage());
+        }
```

Comment:
I don't understand the need for all this complexity here. If the ALTER table command results in an error, the only thing that matters is that something went wrong and we will rollback. What exactly went wrong it not that important (I think). Why are you doing all these checks about the table structure?